quesma/queryparser/pancake_json_rendering.go

```diff
@@ -0,0 +1,160 @@
+		if !isNewBucket {          // for subsequent rows, create new bucket if any key is different
+			previousBucket := buckets[len(buckets)-1]
+			for _, cols := range row.Cols {
+				if strings.HasPrefix(cols.ColName, bucketKeyName) {
+					for _, previousCols := range previousBucket.Cols {
+						if cols.ColName == previousCols.ColName {
+							if cols.Value != previousCols.Value {
+								isNewBucket = true
+							}
+							break
+						}
+					}
+				}
+			}
+		}
+
+		if isNewBucket {
+			buckets = append(buckets, model.QueryResultRow{Index: indexName})
+			subAggrs = append(subAggrs, []model.QueryResultRow{})
+			lastIdx := len(buckets) - 1
+			for _, cols := range row.Cols {
+				if strings.HasPrefix(cols.ColName, bucketKeyName) || strings.HasPrefix(cols.ColName, bucketCountName) {
+					buckets[lastIdx].Cols = append(buckets[lastIdx].Cols, cols)
+				}
+			}
+		}
+		lastIdx := len(buckets) - 1
+		subAggrs[lastIdx] = append(subAggrs[lastIdx], row)
+	}
+
+	return buckets, subAggrs
+}
+
+// In some queries we want to filter out null values or empty
+// We accomplish that by increasing limit by one during SQL query and then filtering out during JSON rendering.
+// So we either filter out empty or last one if there is none.
+// This can't be replaced by WHERE in generic case.
+func (p *pancakeJSONRenderer) potentiallyRemoveExtraBucket(layer *pancakeAggregationLayer, bucketName string, bucketRows []model.QueryResultRow, subAggrRows [][]model.QueryResultRow) ([]model.QueryResultRow, [][]model.QueryResultRow) {
+	// We are filter out null
+	if layer.nextBucketAggregation.filterOurEmptyKeyBucket {
+		nullRowToDelete := -1
+		bucketKeyName := bucketName + "key"
+	ROW:
+		for i, row := range bucketRows {
+			for _, col := range row.Cols {
+				if strings.HasPrefix(col.ColName, bucketKeyName) {
+					if col.Value == nil || col.Value == "" { // TODO: replace with schema
+						nullRowToDelete = i
+						break ROW
+					}
+				}
+			}
+		}
+
+		if nullRowToDelete != -1 {
+			bucketRows = append(bucketRows[:nullRowToDelete], bucketRows[nullRowToDelete+1:]...)
+			subAggrRows = append(subAggrRows[:nullRowToDelete], subAggrRows[nullRowToDelete+1:]...)
+		} else if layer.nextBucketAggregation.limit != 0 && len(bucketRows) > layer.nextBucketAggregation.limit {
+			bucketRows = bucketRows[:layer.nextBucketAggregation.limit]
+			subAggrRows = subAggrRows[:layer.nextBucketAggregation.limit]
+		}
+	}
+	return bucketRows, subAggrRows
+}
+
+func (p *pancakeJSONRenderer) layerToJSON(layerIdx int, layers []*pancakeAggregationLayer, rows []model.QueryResultRow) (model.JsonMap, error) {
+
+	result := model.JsonMap{}
+	if layerIdx >= len(layers) {
+		return result, nil
+	}
+	layer := layers[layerIdx]
+	for _, metric := range layer.currentMetricAggregations {
+		metricName := ""
+		for _, prevLayer := range layers[:layerIdx] {
+			metricName = fmt.Sprintf("%s%s__", metricName, prevLayer.nextBucketAggregation.name)
```

Comment:
Maybe we should store these mangled names in the `pancakeAggregationLayer`.  Right now, the contract between the generator and renderer is fragile.

Reply:
I agree; it's easier to store or have common method.